client/api_object.go

```diff
@@ -382,6 +558,118 @@ func (c *client) FGetObject(ctx context.Context, bucketName, objectName, filePat
 	return nil
 }
 
+// GetSegmentEnd calculates the end position
+func GetSegmentEnd(begin int64, total int64, per int64) int64 {
+	if begin+per > total {
+		return total - 1
+	}
+	return begin + per - 1
+}
+
+// FGetObjectResumable download s3 object payload and return the related object info
+func (c *client) FGetObjectResumable(ctx context.Context, bucketName, objectName, filePath string, opts types.GetObjectOption) error {
+	tempFilePath := filePath + types.TempFileSuffix
+
+	var (
+		startOffset  int64
+		segmentSize  int64
+		objectOption types.GetObjectOption
+		segNum       int64
+	)
+
+	params, err := c.GetParams()
+	if err != nil {
+		return err
+	}
+	segmentSize = int64(params.GetMaxSegmentSize())
+
+	fileInfo, err := os.Stat(tempFilePath)
+	if err != nil {
+		if strings.Contains(err.Error(), "no such file or directory") {
+			startOffset = 0
+		} else {
+			return err
+		}
+	} else {
+		fileSize := fileInfo.Size()
+		startOffset = (fileSize / segmentSize) * segmentSize
+
+		// truncated file to segment size integer multiples
+		if uint64(fileSize)%params.GetMaxSegmentSize() != 0 {
+			file, err := os.OpenFile(tempFilePath, os.O_RDWR, 0644)
+			if err != nil {
+				return err
+			}
+			defer file.Close()
+
+			err = file.Truncate(startOffset)
+			if err != nil {
+				return err
+			}
+			log.Debug().Msgf("The file was truncated to the specified size.%d\n", startOffset)
+			// TODO(chris): verify file's segment
+		}
+	}
+
+	// Get the object detailed meta for object whole size
+	// must delete header:range to get whole object size
+	meta, err := c.HeadObject(ctx, bucketName, objectName)
+	if err != nil {
+		return err
+	}
+
+	// Create the file if not exists. Otherwise the segments download will overwrite it.
+	fd, err := os.OpenFile(tempFilePath, os.O_WRONLY|os.O_CREATE|os.O_APPEND, types.FilePermMode)
+	if err != nil {
+		return err
+	}
+	_, err = fd.Seek(startOffset, io.SeekStart)
+	if err != nil {
+		fd.Close()
+		return err
+	}
+
+	segNum = startOffset / segmentSize
+	// Wait for the segments download finished
+	for offset := startOffset; offset < int64(meta.PayloadSize); offset += segmentSize {
+		// hook for test
+		if err = DownloadSegmentHooker(segNum); err != nil {
+			return err
+		}
+
+		endOffset := GetSegmentEnd(offset, int64(meta.PayloadSize), segmentSize)
```

Comment:
The first and last download segment may be smaller than the segment size. 

```

| -----seg1 ------ | -----seg2 ------ | -----seg3 ------ | -----seg4 ------ |
        ^
        |
 start offset
```
So the GetSegmentEnd may be incorrect.